src/lib/hash/streebog/streebog.cpp

```diff
@@ -0,0 +1,165 @@
+/*
+* Streebog
+* (C) 2017 Ribose Inc.
+*
+* Botan is released under the Simplified BSD License (see license.txt)
+*/
+
+#include <botan/streebog.h>
+#include <botan/exceptn.h>
+
+namespace Botan {
+
+std::unique_ptr<HashFunction> Streebog::copy_state() const
+   {
+   return std::unique_ptr<HashFunction>(new Streebog(*this));
+   }
+
+namespace {
+
+static inline void addm(const uint8_t* m, uint64_t* h)
+   {
+   bool carry = false, overflow = false;
+   for(int i = 0; i < 8; i++)
+      {
+      const uint64_t m64 = load_le<uint64_t>(m, i);
+      const uint64_t hi = load_le<uint64_t>(reinterpret_cast<uint8_t*>(h), i);
+      uint64_t t = hi + m64;
+
+      overflow = t < hi || t < m64;
+      store_le(t + carry, reinterpret_cast<uint8_t*>(&h[i]));
+      carry = overflow;
+      }
+   }
+
+inline void lps(uint64_t* block)
+   {
+   uint8_t r[64];
+   std::memcpy(r, block, 64);
+
+   for(int i = 0; i < 8; ++i)
+      {
+      block[i] =  STREEBOG_Ax[0][r[i]] ^
+                  STREEBOG_Ax[1][r[i + 8]] ^
+                  STREEBOG_Ax[2][r[i + 16]] ^
+                  STREEBOG_Ax[3][r[i + 24]] ^
+                  STREEBOG_Ax[4][r[i + 32]] ^
+                  STREEBOG_Ax[5][r[i + 40]] ^
+                  STREEBOG_Ax[6][r[i + 48]] ^
+                  STREEBOG_Ax[7][r[i + 56]];
+      }
+   }
+
+inline void e(uint64_t* K, const uint8_t* m)
+   {
+   uint64_t tmp[8];
```

Comment:
It is confusing here how `tmp` is used to refer to what is (in RFC 6986) called `K`, and `K` is what is called `m` in the RFC. Can we rename them in some way that will make the usage more clear?